File: Conq/Competição 2022/Soluções/H - Haja Interruptores!/h.cpp
/*
  Problema H - Haja interruptores
  Solução gulosa, seguindo a estratégia descrita no enunciado, pode-se
  perceber que ao final de uma segunda passada pelos interruptores ele
  retornará ao estado inicial, entrando em um ciclo. Sendo assim, a solução
  consiste em realizar estas duas passadas, procurando uma situação em que
  todas as lãmpadas estejam apagadas. Caso essa situação não ocorra, então
  é porque o problema não tem solução.
*/
#include<bits/stdc++.h>

using namespace std;

int qtdeInt, qtdeLamp, qtdeLampAcesas;
bool lampadaAcesa[1000];
int interruptores[1000][1000];
int lampAfetadas[1000];

int resolver()
{
	int i, j, pos;
	int qtdeLampApagadas = qtdeLamp - qtdeLampAcesas; //Qtde de lâmpadas apagadas no começo
	int mov = 0;
	
	if (qtdeLampApagadas == qtdeLamp) //Caso em que o estado inicial já é a solução
		return mov;
		
	for(i = 0; i < qtdeInt; i++) //Primeira passada
	{
		mov++;
		for(j = 0; j < lampAfetadas[i]; j++){ //Acionando interruptores
			pos = interruptores[i][j];
			lampadaAcesa[pos] = !lampadaAcesa[pos];
			if (!lampadaAcesa[pos]) //Atualizando número de lâmpadas apagadas
				qtdeLampApagadas++;
			else
				qtdeLampApagadas--;
		}
		if (qtdeLampApagadas == qtdeLamp) //Se todas as lâmpadas estiverem apagadas, é a solução
			return mov;
	}
	for(i = 0; i < qtdeInt; i++){ //Segunda passada
		mov++;
		for(j = 0; j < lampAfetadas[i]; j++){
			pos = interruptores[i][j];
			lampadaAcesa[pos] = !lampadaAcesa[pos];
			if (!lampadaAcesa[pos])
				qtdeLampApagadas++;
			else
				qtdeLampApagadas--;
		}
		if (qtdeLampApagadas == qtdeLamp)
			return mov;
	}
	//Se chegar aqui, é pq não encontramos solução
	return -1;
}

int main(){
	int i, j;
	
	scanf("%d %d", &qtdeInt, &qtdeLamp);
	for(i = 0; i < qtdeLamp; i++)
		lampadaAcesa[i] = false;
	
	scanf("%d", &qtdeLampAcesas);
	
	for(i = 0; i < qtdeLampAcesas; i++)
	{
		scanf("%d", &j);
		lampadaAcesa[j-1] = true;
	}
	
	for(i = 0; i < qtdeInt; i++)
	{
		scanf("%d", &lampAfetadas[i]);
		for(j = 0; j < lampAfetadas[i]; j++)
		{
			scanf("%d", &interruptores[i][j]);
			interruptores[i][j]--;
		}
	}
	printf("%d\n",resolver());
}

File: Conq/Competição 2022/Soluções/H - Haja Interruptores!/h.py
def resolver(lampAcesas, interruptores, qtdeLampAcesas):
    qtdeInt = len(interruptores)
    qtdeLamp = len(lampAcesas)
    qtdeLampApagadas = qtdeLamp - qtdeLampAcesas

    if qtdeLampApagadas == qtdeLamp:
        return 0

    mov = 0
    for k in range(2):
        for i in range(qtdeInt):
            mov += 1
            for pos in interruptores[i]:
                lampAcesas[pos] = not lampAcesas[pos]
                if lampAcesas[pos]:
                    qtdeLampApagadas -= 1
                else:
                    qtdeLampApagadas += 1
            if qtdeLampApagadas == qtdeLamp:
                return mov
    return -1

qtdeInt, qtdeLamp = [int(x) for x in input().split()]

entrada = [int(x) for x in input().split()]
qtdeLampAcesas = entrada[0]
lampAcesa = [False for x in range(qtdeLamp)]
for i in range(qtdeLampAcesas):
    lampAcesa[entrada[i+1]-1] = True

interruptores = []
for i in range(qtdeInt):
    entrada = [int(x)-1 for x in input().split()]
    interruptores.append(entrada[1:])
print(resolver(lampAcesa, interruptores, qtdeLampAcesas))